### PR TITLE
refactor remote file handline

### PR DESCRIPTION
**Pull request by nevillelyh**

### CODE REVIEW

File: scio-core/src/main/java/com/spotify/scio/util/RemoteFileUtil.java [hunk 1]
```diff
@@ -0,0 +1,320 @@
+/*
```
Comment (andrewsmartin) [inline scio-core/src/main/java/com/spotify/scio/util/RemoteFileUtil.java]: Why Java?

Comment (nevillelyh) [inline scio-core/src/main/java/com/spotify/scio/util/RemoteFileUtil.java]: It's mostly using Java APIs. And this class maybe reused in a pure Java DF pipeline.

File: scio-core/src/main/java/com/spotify/scio/util/RemoteFileUtil.java [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+            shouldDownload = false;
+          } else {
+            LOG.warn("Destination exists with wrong size. {} [{}] -> {} [{}]",
+                src, srcSize, dst, dstSize);
+            Files.delete(dst);
+          }
+        }
+        if (shouldDownload) {
+          copyToLocal(src, dst);
+          LOG.info("Downloaded {} -> {} [{}]", src, dst, srcSize);
```
Comment (ravwojdyla) [inline scio-core/src/main/java/com/spotify/scio/util/RemoteFileUtil.java]: `"Downloaded {} -> {} [{}] bytes"`?

File: scio-core/src/main/java/com/spotify/scio/util/RemoteFileUtil.java [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+        // Remote URI
+        long srcSize = getRemoteSize(src);
+        boolean shouldDownload = true;
+        if (Files.exists(dst)) {
+          long dstSize = Files.size(dst);
+          if (srcSize == dstSize) {
+            LOG.info("URI {} already downloaded", src);
+            shouldDownload = false;
+          } else {
+            LOG.warn("Destination exists with wrong size. {} [{}] -> {} [{}]",
```
Comment (ravwojdyla) [inline scio-core/src/main/java/com/spotify/scio/util/RemoteFileUtil.java]: "[{}] bytes"

**Comment by ravwojdyla** [comment]
Tests are missing for the new logic.

File: scio-extra/src/main/scala/com/spotify/scio/extra/sparkey/SparkeyUri.scala [hunk 4]
```diff
@@ -49,55 +44,43 @@ trait SparkeyUri extends Serializable {
...
+      .exists(e => rfu.remoteExists(new URI(basePath + e)))
 }
 
 private[sparkey] class SparkeyWriter(val uri: SparkeyUri) {
 
   private val localFile = uri match {
-    case u: LocalSparkeyUri => u.toString
-    case u: GcsSparkeyUri => u.localBasePath
+    case u: LocalSparkeyUri => u.basePath
+    case _: RemoteSparkeyUri => Files.createTempDirectory("sparkey").resolve("data").toString
```
Comment (andrewsmartin) [inline scio-extra/src/main/scala/com/spotify/scio/extra/sparkey/SparkeyUri.scala]: So the hash prefix is no longer needed right? This should be unique every time it's called / internally uses some kind of hash in the name?

Comment (andrewsmartin) [inline scio-extra/src/main/scala/com/spotify/scio/extra/sparkey/SparkeyUri.scala]: Yeah uses random numbers 👍

File: scio-core/src/main/java/com/spotify/scio/util/RemoteFileUtil.java (hunk 2 above)
Comment (ravwojdyla) [inline scio-core/src/main/java/com/spotify/scio/util/RemoteFileUtil.java]: `B` sounds good and simple.

**Pull request merged**